app/controllers/users_email_controller.rb

```diff
@@ -29,38 +47,142 @@ def update
+    if @error
+      flash[:error] = @error
+      redirect_to redirect_url
+    else
+      redirect_to "#{redirect_url}?done=true"
+    end
+  end
+
+  def show_confirm_new_email
+    load_change_request(:new)
 
-      @update_result = :invalid_second_factor
-      @backup_codes_enabled = true if user.backup_codes_enabled?
+    if params[:done].to_s == "true"
+      @done = true
+    end
+
+    if @change_request&.change_state != EmailChangeRequest.states[:authorizing_new]
+      @error = I18n.t("change_email.already_done")
+    end
 
-      if params[:second_factor_token].present?
-        RateLimiter.new(nil, "second-factor-min-#{request.remote_ip}", 3, 1.minute).performed!
-        @show_invalid_second_factor_error = true
+    @show_invalid_second_factor_error = flash[:invalid_second_factor]
+
+    if !@error
+      if @user.totp_enabled?
+        @backup_codes_enabled = @user.backup_codes_enabled?
+        if params[:show_backup].to_s == "true" && @backup_codes_enabled
+          @show_backup_codes = true
+        else
+          @show_second_factor = true
+        end
       end
-    else
-      updater = EmailUpdater.new
-      @update_result = updater.confirm(params[:token])
 
-      if @update_result == :complete
-        updater.user.user_stat.reset_bounce_score!
-        log_on_user(updater.user)
+      @to_email = @change_request.new_email
+    end
+
+    render layout: 'no_ember'
+  end
+
+  def confirm_old_email
+    load_change_request(:old)
+
+    if @change_request&.change_state != EmailChangeRequest.states[:authorizing_old]
+      @error = I18n.t("change_email.already_done")
+    end
+
+    redirect_url = path("/u/confirm-old-email/#{params[:token]}")
+
+    if !@error
+      updater = EmailUpdater.new
+      if updater.confirm(params[:token]) != :authorizing_new
+        @error = I18n.t("change_email.already_done")
       end
     end
 
+    if @error
+      flash[:error] = @error
+      redirect_to redirect_url
+    else
+      redirect_to "#{redirect_url}?done=true"
+    end
+  end
+
+  def show_confirm_old_email
+    load_change_request(:old)
+
+    if @change_request&.change_state != EmailChangeRequest.states[:authorizing_old]
+      @error = I18n.t("change_email.already_done")
+    end
+
+    if params[:done].to_s == "true"
+      @almost_done = true
+    end
+
+    if !@error
+      @from_email = @user.email
+      @to_email = @change_request.new_email
+    end
+
     render layout: 'no_ember'
   end
 
+  private
+
+  def load_change_request(type)
+    expires_now
+
+    @token = EmailToken.confirmable(params[:token])
+
+    if @token
+      if type == :old
+        @change_request = @token.user&.email_change_requests.where(old_email_token_id: @token.id).first
+      elsif type == :new
+        @change_request = @token.user&.email_change_requests.where(new_email_token_id: @token.id).first
+      end
+    end
+
+    if (!@token || !@token&.user || !@change_request)
```

Comment:
```suggestion
    if (!@token&.user || !@change_request)
```

No need to check for token if you use the `&` operator.